src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -164,82 +170,95 @@ public Assembly LoadFromPath(string originalAnalyzerPath)
 
             lock (_guard)
             {
-                _analyzerAssemblyInfoMap[originalAnalyzerPath] = (assemblyName, realPath, resourceAssemblyCultureNames);
+                _analyzerAssemblyInfoMap[originalAnalyzerPath] = (assemblyName, realPath);
             }
 
-            return (assemblyName, realPath, resourceAssemblyCultureNames);
+            return (assemblyName, realPath);
+        }
+
+        internal string? GetSatelliteInfoForPath(string originalAnalyzerPath, CultureInfo cultureInfo)
+        {
+            Dictionary<CultureInfo, string?>? satelliteDictionary;
+            string? satelliteAssemblyPath = null;
 
-            // Discover the culture names for any satellite dlls related to this analyzer. These 
-            // need to be understood when handling the resource loading in certain cases.
-            static ImmutableHashSet<string> getResourceAssemblyCultureNames(string originalAnalyzerPath)
+            lock (_guard)
             {
-                var path = Path.GetDirectoryName(originalAnalyzerPath)!;
-                using var enumerator = Directory.EnumerateDirectories(path, "*").GetEnumerator();
-                if (!enumerator.MoveNext())
+                if (_analyzerSatelliteAssemblyInfoMap.TryGetValue(originalAnalyzerPath, out satelliteDictionary))
                 {
-                    return ImmutableHashSet<string>.Empty;
+                    if (satelliteDictionary.TryGetValue(cultureInfo, out satelliteAssemblyPath))
+                    {
+                        return satelliteAssemblyPath;
+                    }
+                }
+                else
+                {
+                    satelliteDictionary = new Dictionary<CultureInfo, string?>();
+                    _analyzerSatelliteAssemblyInfoMap[originalAnalyzerPath] = satelliteDictionary;
                 }
+            }
+
+            (satelliteAssemblyPath, var actualCultureName) = getSatelliteAssemblyLocation(originalAnalyzerPath, cultureInfo);
+            if (actualCultureName != null)
+            {
+                PrepareSatelliteAssemblyToLoad(originalAnalyzerPath, actualCultureName);
```

Comment:
In this model it's possible that `PrepareSatelliteAssemblyToLoad` is called twice with the same arguments. Consider if only the `ES` culture assembly is present but we get calls to load `es-es` then `es`. Will the impl of this method be correct in the case it's called twice with same arguments?

Reply:
I *think* the current code is fine. This code mimics how the non-satellite assembly is prepared via the call to PreparePathToLoad, as that could also be called twice with the same arg. The walk up the CultureInfo chain happens on the original path, which shouldn't be affected by any simultaneous file io occurring via PrepareSatelliteAssemblyToLoad.